resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/Folder.java

```diff
@@ -47,9 +48,10 @@ protected void assertCreateRequirements(IFileStore store, IFileInfo localInfo, i
 		if (!force && localInfo.exists()) {
 			//return an appropriate error message for case variant collisions
 			if (!Workspace.caseSensitive) {
-				String name = getLocalManager().getLocalName(store);
+				String name = localInfo.getName();
 				if (name != null && !store.getName().equals(name)) {
-					String msg = NLS.bind(Messages.resources_existsLocalDifferentCase, IPath.fromOSString(store.toString()).removeLastSegments(1).append(name).toOSString());
+					String msg = NLS.bind(Messages.resources_existsLocalDifferentCase,
+							Path.of(store.toString()).resolveSibling(name));
 					throw new ResourceException(IResourceStatus.CASE_VARIANT_EXISTS, getFullPath(), msg, null);
```

Comment:
CASE_VARIANT_EXISTS makes only sense when name is tested against the actual name in the filesystem. This codepath is only taken if localInfo.exists(), so never in the performance critical case where where the directory is clean.